packages/forklift-console-plugin/src/modules/Providers/views/migrate/reducer.ts

```diff
@@ -0,0 +1,153 @@
+
+const actions: {
+  [name: string]: (
+    draft: Draft<CreateVmMigrationPageState>,
+    action: PageAction<CreateVmMigration, unknown>,
+  ) => CreateVmMigrationPageState;
+} = {
+  [SET_NAME](
+    draft,
+    { payload: { name, existingPlanNames } }: PageAction<CreateVmMigration, PlanName>,
+  ) {
+    draft.newPlan.metadata.name = name;
+    draft.validation.name =
+      validateK8sName(name) && validateUniqueName(name, existingPlanNames) ? 'success' : 'error';
+    return draft;
+  },
+  [SET_DESCRIPTION](
+    draft,
+    { payload: { description } }: PageAction<CreateVmMigration, PlanDescription>,
+  ) {
+    draft.newPlan.spec.description = description;
+    return draft;
+  },
+  [SET_TARGET_NAMESPACE](
+    draft,
+    { payload: { targetNamespace } }: PageAction<CreateVmMigration, PlanTargetNamespace>,
+  ) {
+    draft.newPlan.spec.targetNamespace = targetNamespace;
+    draft.validation.targetNamespace = validateK8sName(targetNamespace) ? 'success' : 'error';
+    return draft;
+  },
+  [SET_TARGET_PROVIDER](
+    draft,
+    { payload: { targetProvider } }: PageAction<CreateVmMigration, PlanTargetProvider>,
+  ) {
+    draft.newPlan.spec.provider.destination = getObjectRef(targetProvider);
+    draft.newPlan.spec.targetNamespace = undefined;
+    draft.validation.targetNamespace = 'default';
+    return draft;
+  },
+  [SET_AVAILABLE_PROVIDERS](
+    draft,
+    { payload: { availableProviders } }: PageAction<CreateVmMigration, PlanAvailableProviders>,
+  ) {
+    const targetProvider = draft.newPlan.spec.provider.destination;
+    if (
+      !targetProvider ||
+      !availableProviders.find((p) => p?.metadata?.name === targetProvider.name)
+    ) {
+      // set the default provider if none is set
+      // reset the provider if provider was removed
+      const firstHostProvider = availableProviders.find((p) => isProviderLocalTarget(p));
+      draft.newPlan.spec.provider.destination =
+        firstHostProvider && getObjectRef(firstHostProvider);
+      draft.newPlan.spec.targetNamespace = undefined;
+      draft.validation.targetNamespace = 'default';
+    }
+    draft.availableProviders = availableProviders;
+    return draft;
+  },
+};
+
+export const reducer = (
+  draft: Draft<CreateVmMigrationPageState>,
+  action: PageAction<CreateVmMigration, unknown>,
+) => {
+  return actions?.[action?.type]?.(draft, action) ?? draft;
+};
+
+// based on the method used in legacy/src/common/helpers
+// and mocks/src/definitions/utils
+export const getObjectRef = (
+  { apiVersion, kind, metadata: { name, namespace, uid } = {} }: V1beta1Provider = {
+    apiVersion: undefined,
+    kind: undefined,
+  },
+) => ({
+  apiVersion,
+  kind,
+  name,
+  namespace,
+  uid,
+});
+
+export const createInitialState = ({
+  namespace,
+  sourceProvider,
+  selectedVmIds,
+}: {
+  namespace: string;
+  sourceProvider: V1beta1Provider;
+  selectedVmIds: string[];
+}): CreateVmMigrationPageState => ({
+  newPlan: {
+    ...planTemplate,
```

Comment:
do we get new information from the global const `planTemplate`, can we inline the information ? is `planTemplate` mutable ? will we get different information in different runs ?

Reply:
hmm, ok
i prefer inline in this case, because I think it's very little new info in `planTemplate`, just `kind` and `version` ?
but I do not care that much ... maybe we will add more into some day ?